darts/timeseries.py

```diff
@@ -3,117 +3,554 @@
 ----------
 
 `TimeSeries` is the main class in `darts`. It represents a univariate or multivariate time series.
+It can represent a stochastic time series by storing several samples (trajectories).
 """
 
 import pandas as pd
 import numpy as np
-from copy import deepcopy
+import xarray as xr
 import matplotlib.pyplot as plt
-from pandas.tseries.frequencies import to_offset
 from typing import Tuple, Optional, Callable, Any, List, Union
 from inspect import signature
+from collections import defaultdict
+from pandas.tseries.frequencies import to_offset
 
 from .logging import raise_log, raise_if_not, raise_if, get_logger
 
 logger = get_logger(__name__)
 
+# dimension names in the DataArray
+# the "time" one can be different, if it has a name in the underlying Series/DataFrame.
+DIMS = ('time', 'component', 'sample')
+
 
 class TimeSeries:
-    def __init__(self,
-                 df: pd.DataFrame,
-                 freq: Optional[str] = None,
-                 fill_missing_dates: Optional[bool] = True,
-                 dummy_index: Optional[bool] = False):
+    def __init__(self, xa: xr.DataArray):
+        """
+        Wrapper around a (well formed) DataArray. Use the static factory methods to build instances unless
+        you know what you are doing.
```

Comment:
A few replies to your points:
* Overall, the approach I'm trying to follow here is really that a TimeSeries is uniquely determined by the DataArray it's wrapping around, in an immutable way. In particular, I don't think it should be determined by the combination of (DataArray + "filling the dates on this array" + freq, ... etc). This is because doing this IMO makes it harder to reason about TimeSeries because we have to reason both about the data it contains (i.e., the DataFrame or DataArray), and also about the processing that might or might not have been done on this data. I think it's a better practice (and more functional) to consider the data immutable and split the "data wrapping" part (TimeSeries) from the processing part (factory methods). This is (somewhat) inspired from "case classes" in Scala for instance.
* Indeed here we are sacrificing backward compatibility of the constructor. But I don't think there's a way around it if we make the choice to wrap around DataArrays instead of DataFrames.
* Also I don't think there's really much of a cost for users to be calling factory methods most of the time instead of the constructor. We were already doing that in most of our examples before. Factory methods are great! They allow to do translations between whatever format the users want to use (pandas, numpy, xarray, json, ...) and correct DataArray representations, and encapsulate all the processing logic.